File: 프로그래머스/2/388353. 지게차와 크레인/지게차와 크레인.java
import java.util.*;

class Solution {
    int n, m;
    char[][] grid;
    int[] dx = {0, 0, 1, -1};
    int[] dy = {1, -1, 0, 0};

    public int solution(String[] storage, String[] requests) {
        n = storage.length;
        m = storage[0].length();
        grid = new char[n + 2][m + 2];

        for (char[] row : grid) Arrays.fill(row, '.');

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                grid[i + 1][j + 1] = storage[i].charAt(j);
            }
        }

        for (String request : requests) {
            char target = request.charAt(0);
            if (request.length() == 2) {
                for (int i = 1; i <= n; i++) {
                    for (int j = 1; j <= m; j++) {
                        if (grid[i][j] == target) grid[i][j] = '#';
                    }
                }
            } else {
                boolean removed = true;
                while (removed) {
                    removed = false;
                    for (int i = 1; i <= n; i++) {
                        for (int j = 1; j <= m; j++) {
                            if (grid[i][j] == target) {
                                for (int dir = 0; dir < 4; dir++) {
                                    int nx = i + dx[dir];
                                    int ny = j + dy[dir];
                                    if (grid[nx][ny] == '.') {
                                        grid[i][j] = '#';
                                        removed = true;
                                        break;
                                    }
                                }
                            }
                        }
                    }
                }
            }
            
            bfs();
        }

        int count = 0;
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                if (grid[i][j] != '.' && grid[i][j] != '#') count++;
            }
        }

        return count;
    }

    private void bfs() {
        Queue<int[]> queue = new ArrayDeque<>();
        for (int i = 0; i < n + 2; i++) {
            for (int j = 0; j < m + 2; j++) {
                if (grid[i][j] == '.') queue.add(new int[]{i, j});
            }
        }

        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            for (int dir = 0; dir < 4; dir++) {
                int nx = cur[0] + dx[dir];
                int ny = cur[1] + dy[dir];
                if (nx < 0 || nx >= n + 2 || ny < 0 || ny >= m + 2) continue;
                if (grid[nx][ny] == '#') {
                    grid[nx][ny] = '.';
                    queue.offer(new int[]{nx, ny});
                }
            }
        }
    }
}

File: 프로그래머스/2/388353. 지게차와 크레인/지게차와 크레인.kt
import java.util.ArrayDeque

class Solution {
    private var n = 0
    private var m = 0
    private lateinit var grid: Array<CharArray>
    private val dx = intArrayOf(0, 0, 1, -1)
    private val dy = intArrayOf(1, -1, 0, 0)

    fun solution(storage: Array<String>, requests: Array<String>): Int {
        n = storage.size
        m = storage[0].length
        grid = Array(n + 2) { CharArray(m + 2) { '.' } }

        for (i in 0 until n)
            for (j in 0 until m)
                grid[i + 1][j + 1] = storage[i][j]

        for (request in requests) {
            val target = request[0]
            if (request.length == 2) {
                for (i in 1..n)
                    for (j in 1..m)
                        if (grid[i][j] == target) grid[i][j] = '#'
            } else {
                var removed = true
                while (removed) {
                    removed = false
                    for (i in 1..n) {
                        for (j in 1..m) {
                            if (grid[i][j] == target) {
                                for (dir in 0..3) {
                                    if (grid[i + dx[dir]][j + dy[dir]] == '.') {
                                        grid[i][j] = '#'
                                        removed = true
                                        break
                                    }
                                }
                            }
                        }
                    }
                }
            }
            bfs()
        }

        return (1..n).sumOf { i ->
            (1..m).count { j -> grid[i][j] != '.' && grid[i][j] != '#' }
        }
    }

    private fun bfs() {
        val queue = ArrayDeque<IntArray>()
        for (i in 0 until n + 2)
            for (j in 0 until m + 2)
                if (grid[i][j] == '.') queue.add(intArrayOf(i, j))

        while (queue.isNotEmpty()) {
            val (x, y) = queue.poll()
            for (dir in 0..3) {
                val nx = x + dx[dir]
                val ny = y + dy[dir]
                if (nx < 0 || nx >= n + 2 || ny < 0 || ny >= m + 2) continue
                if (grid[nx][ny] == '#') {
                    grid[nx][ny] = '.'
                    queue.add(intArrayOf(nx, ny))
                }
            }
        }
    }
}